src/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -210,269 +211,245 @@ public sealed partial class TokenCache : ITokenCacheInternal
         }
 
         async Task<bool?> ITokenCacheInternal.IsFociMemberAsync(AuthenticationRequestParameters requestParams, string familyId)
         {
             var logger = requestParams.RequestContext.Logger;
             if (requestParams?.AuthorityInfo?.CanonicalAuthority == null)
             {
-                logger.Warning("No authority details, can't check app metadta. Returning unkown");
+                logger.Warning("No authority details, can't check app metadata. Returning unknown");
                 return null;
             }
 
-            InstanceDiscoveryMetadataEntry instanceDiscoveryMetadata =
-                   await ServiceBundle.InstanceDiscoveryManager.GetMetadataEntryAsync(
-                       new Uri(requestParams.AuthorityInfo.CanonicalAuthority),
-                       requestParams.RequestContext)
-                   .ConfigureAwait(false);
-
-            TokenCacheNotificationArgs args = new TokenCacheNotificationArgs(this, ClientId, requestParams?.Account, false);
+            IEnumerable<MsalAppMetadataCacheItem> allAppMetadata = Enumerable.Empty<MsalAppMetadataCacheItem>();
+            await LoadFromCacheAsync(
+                    requestParams.RequestContext.TelemetryCorrelationId,
+                    CacheEvent.TokenTypes.APPMETADATA,
+                    requestParams.Account,
+                    () => allAppMetadata = _accessor.GetAllAppMetadata())
+                .ConfigureAwait(false);
 
-            //TODO: bogavril - is the env ok here? Can I cache it or pass it in?
-            MsalAppMetadataCacheItem appMetadata;
-            await _semaphoreSlim.WaitAsync().ConfigureAwait(false);
-            try
-            {
-                await OnBeforeAccessAsync(args).ConfigureAwait(false);
+            var instanceMetadata = await ServiceBundle.InstanceDiscoveryManager.GetMetadataEntryTryAvoidNetworkAsync(
+                    requestParams.AuthorityInfo.CanonicalAuthority,
+                    allAppMetadata.Select(m => m.Environment),
+                    requestParams.RequestContext)
+                .ConfigureAwait(false);
 
-                appMetadata =
-                    instanceDiscoveryMetadata.Aliases
-                    .Select(env => _accessor.GetAppMetadata(new MsalAppMetadataCacheKey(ClientId, env)))
-                    .FirstOrDefault(item => item != null);
-
-                await OnAfterAccessAsync(args).ConfigureAwait(false);
-            }
-            finally
-            {
-                _semaphoreSlim.Release();
-            }
+            var appMetadata =
+                instanceMetadata.Aliases
+                .Select(env => _accessor.GetAppMetadata(new MsalAppMetadataCacheKey(ClientId, env)))
+                .FirstOrDefault(item => item != null);
 
             if (appMetadata == null)
             {
-                logger.Warning("No app metadata found. Returning unkown");
+                logger.Warning("No app metadata found. Returning unknown");
```

Comment:
>unknown"); [](start = 65, length = 10)

what is unknown here?

Reply:
It's a FOCI status. An app (client id) can have 3 states - in the family, not in the family or unkown. In hindsight, I should have created an enum for this.

Reply:
could you add something more to the comment, so the "Unknown" part is clear. Just thinking if we get an icm or logs from a customer and someone see this, who isn't you, it's not clear from the log message.

Reply:
#supportability

---
In reply to: [298342083](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1229#discussion_r298342083) [](ancestors = 298342083)